### ISSUE TITLE
Add support for SBOM analyzing at Binary-Artifacts stage by integrating `sbom-scorecard`

### ISSUE THREAD

**Issue by Dentrax** [ossf/scorecard#2605]
**Is your feature request related to a problem? Please describe.**
We should [Detect if SBOMs generated](https://github.com/ossf/scorecard/issues/1476) (by @david-a-wheeler), and then we can scan the SBOMs to enrich the overall scorecard score using 3rd-party packages.

P.S.: `scan` in this context, does not mean _vulnerability scanning_. We can think of it like _SBOM best practices & compliance scanning_.

**Describe the solution you'd like**
Integrate the [sbom-scorecard](https://github.com/eBay/sbom-scorecard) for calculating SBOM scores and aggregate the score result to `Binary-Artifacts` score.

See the [getScore()](https://github.com/eBay/sbom-scorecard/blob/b324861a7e2c6d538143bd4f07bd05d2b70059df/pkg/scorecard/scorecard.go#L57) function for more information.

**Describe alternatives you've considered**
Do not scan SBOMs if it's out-of-context of scorecard.

**Additional context**
`-`

/cc @developer-guy @justinabrahms

**Comment by justinabrahms** [comment]
I'm quite happy to help this become a thing. Reach out with any questions you may have. :)

**Comment by laurentsimon** [comment]
Sounds like a good idea. Any chance you could join a bi-weekly Scorecard sync to propose the idea?
In addition to the score, we are working on providing finer-grained results, an example can be found https://github.com/ossf/scorecard/pull/2584, so hopefully SBOM check can be expressed with this format.

**Comment by justinabrahms** [comment]
This is tracked at https://github.com/eBay/sbom-scorecard/issues/34 and likely shouldn't be a blocker for progressing on integration.

**Comment by laurentsimon** [comment]
btw, what do you mean by `Binary-Artifacts stage`?

**Comment by david-a-wheeler** [comment]
>I think @david-a-wheeler's suggestion to review at https://www.ntia.gov/sites/default/files/publications/sbom_minimum_elements_report_0.pdf and ensure these minimum criteria are included in the score. Looks like you are 90% there.

BTW, I don't think this is a blocker for use by OpenSSF *Scorecard*, it was just a suggestion for how to make sbom-scorecard even better. I think it helps if you can point to what some customers are already asking for, so you can say, "yes, we cover that!".

**Comment by Dentrax** [comment]
> btw, what do you mean by `Binary-Artifacts stage`?

Oh, I think I should have said `evaluation` instead: https://github.com/ossf/scorecard/blob/main/checks/evaluation/binary_artifacts.go

**Comment by laurentsimon** [comment]
OK. What's the reasoning behind having the SBOM binaries in the repo (Binary-Artifacts) rather than for released binary assets of the project? 

I'm also wondering whether users will be able to remediate if the SBOM is of poor quality: do existing SBOM generation tools meet the criteria of "good SBOMs"? 

Is a "good SBOM" solely a property of the SBOM generation tool, the project maintainers, or both?

**Comment by justinabrahms** [comment]
I've not looked into whether the tools are capable of generating "good SBOMs" in practice, but there's nothing stopping it technically. It's mostly "have the dependencies added the relevant metadata in the right places". I would say "good sbom" is both the tool and the maintainers.

For my $0.02, I think that you should put the sbom along the artifact, not in source control.

**Comment by Dentrax** [comment]
> What's the reasoning behind checking for SBOM for binaries in the repo (Binary-Artifacts - since we encourage users not to store binaries in the repo) rather than for released binary assets of the project?

>  I think that you should put the sbom along the artifact, not in source control.

SBOMs should NOT be stored in the repo - Probably I dropped the wrong source code - since I'm not familiar with the codebase - so that you thought in this way. SBOMs should be an artifact that served along with the binaries in the release page or smth.

**Comment by justinabrahms** [comment]
In [the meeting today](https://docs.google.com/document/d/1dB2U7_qZpNW96vtuoG7ShmgKXzIg6R5XT5Tc-0yz6kE/edit#heading=h.43s993kl0ppp), we discussed that this should happen within the context of release, not binary artifacts.

We will need a mechanism to detect if something is an sbom given a (for instance) github release. Google's osv-scanner [may provide a way to do that](https://github.com/google/osv-scanner/pull/258). We believe that this will probably look something like looking through the release artifacts for matching files (either as filename searching or attempting to parse the data and see if there's a formatting match).

**Comment by spencerschrock** [comment]
+1 to if Scorecard starts checking for SBOMs, it would fit better with releases, which makes the following checks candidates: `Signed-Releases` and `Packaging`. 

My only reservation is around diluting the score of the `Signed-Releases`. Currently points are awarded as:
8 for signing
2 for provenance

Any points to SBOM would come out of this, and if we wanted more than just a basic "do you have an SBOM?" it would involve multiple such points.

**Comment by justinabrahms** [comment]
I agree that something titled "signed releases" should heavily favor signing. It does seem like there's more to a release than just signing. Is it possible for us to just name this "releases" and then include signing as a subcomponent?

**Comment by lucasgonze** [comment]
I think we’re getting away from the original goal, which is to incorporate SBOMs as a signal, to encourage their generation, and to encourage higher-quality SBOMs. The relevant questions are:

1. Does an SBOM exist?
2. How good is it?
 
If the SBOM is stored in the repo instead of packaged with a release, that is a suboptimal practice, but it doesn’t negate either the existence of the SBOM or its quality.

I propose that a check for question 1 might look for SBOM generation in any detectable form. A check for question 2 then hinges on whether the SBOM is available to be graded.

This is enough complexity to warrant a new section.

**Comment by lucasgonze** [comment]
I researched methods to document SBOM generation using a Github Action:
https://gist.github.com/lucasgonze/b84a2c52c697bf8f686a005080c369dd

Findings:
1. There are a number of SBOM generators that can be deployed using a Github Action. Generating an SBOM in this way is trivial.
2. There is a strong order-of-magnitude popularity curve. The most popular is 128 stars, the next is 28 stars, the rest are below 20. I imagine the most popular (Anchore) accounts for the vast majority of generated SBOMs.
3. Every generator could be detected unambiguously. None of them were multipurpose tools that might or might not be generating an SBOM.
4. Every Action could be detected by examining code under /.github/workflows/
5. Outputs varied widely. Naming conventions were inconsistent. Most allowed output files to be named using a parameter.
6. Only two (the most and least popular, Anchore and Diggity) specifically targeted release artifacts
